File: flipperf/src/main/java/com/flipkart/flipperf/reporter/NetworkEventReporter.java
package com.flipkart.flipperf.reporter;


import java.io.IOException;
import java.net.URL;

import okhttp3.Response;
import okhttp3.ResponseBody;

/**
 * Interface to report events in case of response or any errors.
 */
public interface NetworkEventReporter {

    /**
     * Notifies the {@link NetworkEventReporter} that the intercepted {@link Response} headers has been received
     *
     * @param inspectorResponse : contains response headers
     */
    void responseReceived(InspectorRequest inspectorRequest, InspectorResponse inspectorResponse);

    /**
     * Reports any {@link IOException} while {@link Response} is being proceeded.
     *
     * @param inspectorRequest {@link InspectorRequest}
     * @param e                : error message
     */
    void httpExchangeError(InspectorRequest inspectorRequest, IOException e);

    /**
     * Reports error while getting the input steam from {@link ResponseBody}
     *
     * @param inspectorResponse {@link InspectorResponse}
     * @param e                 : error message
     */
    void responseInputStreamError(InspectorRequest inspectorRequest, InspectorResponse inspectorResponse, Exception e);

    interface InspectorRequest {
        int requestId();

        URL url();

        String method();

        long requestSize();

        String hostName();
    }

    interface InspectorResponse {

        int requestId();

        int statusCode();

        long responseSize();

        long startTime();

        long endTime();
    }
}

File: library-kotlin/src/main/java/com/flipkart/okhttpstats/kotlin/reporter/NetworkEventReporter.kt
package com.flipkart.okhttpstats.kotlin.reporter

import okhttp3.Response
import okhttp3.ResponseBody
import java.io.IOException
import java.net.URL

/**
 * Interface to report events in case of response or any errors.
 */
interface NetworkEventReporter {

    /**
     * Notifies the [NetworkEventReporter] that the intercepted [Response] headers has been received

     * @param inspectorResponse : contains response headers
     */
    fun responseReceived(inspectorRequest: InspectorRequest, inspectorResponse: InspectorResponse)

    /**
     * Reports any [IOException] while [Response] is being proceeded.

     * @param inspectorRequest [InspectorRequest]
     * *
     * @param e                : error message
     */
    fun httpExchangeError(inspectorRequest: InspectorRequest, e: IOException)

    /**
     * Reports error while getting the input steam from [ResponseBody]

     * @param inspectorResponse [InspectorResponse]
     * *
     * @param e                 : error message
     */
    fun responseInputStreamError(inspectorRequest: InspectorRequest, inspectorResponse: InspectorResponse, e: Exception)

    interface InspectorRequest {
        fun requestId(): Int

        fun url(): URL

        fun method(): String

        fun requestSize(): Long

        fun hostName(): String
    }

    interface InspectorResponse {

        fun requestId(): Int

        fun statusCode(): Int

        fun responseSize(): Long

        fun startTime(): Long

        fun endTime(): Long

        fun responseBody(): ResponseBody?
    }
}